### PR TITLE
OCPBUGS#3076 - adding oc mirror info

### PR DESCRIPTION
**Pull request by skopacz1**
Versions 4.11+

[OCPBUGS-3076](https://issues.redhat.com/browse/OCPBUGS-3076)

This PR adds documentation for using oc mirror to mirror images onto a registry when upgrading in a restricted network environment. The new information adds a new section for oc mirror, while keeping the old section describing how to do the same thing with the "adm release mirror" command, which is still supported. The new section points readers to an existing page for disconnected installation with oc mirror, and instructs users to follow the procedures on that page that are relevant to updates.

(some smaller, out of scope changes have also been made to fix incorrect heading levels for adjacent sections)

Preview
Non-OSUS section: https://53575--docspreview.netlify.app/openshift-enterprise/latest/updating/updating-restricted-network-cluster.html#update-mirror-repository_updating-restricted-network-cluster

OSUS section: https://53575--docspreview.netlify.app/openshift-enterprise/latest/updating/updating-restricted-network-cluster.html#update-service-mirror-release_updating-restricted-network-cluster

### CODE REVIEW

File: modules/update-mirror-repository-oc-mirror.adoc [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Module included in the following assemblies:
+//
+// * updating/updating-restricted-network-cluster.adoc
+
+:_content-type: PROCEDURE
+[id="update-mirror-repository_oc_mirror_{context}"]
+= Mirroring images using oc-mirror
+
+Using the oc-mirror OpenShift CLI (`oc`) plug-in is the recommended method for mirroring images onto a mirror registry in a fully or partially disconnected environment.
```
Comment (wking) [inline modules/update-mirror-repository-oc-mirror.adoc:9]: nit: do we have other recommended methods for connected environments?  If not (and I'm not sure what they'd be; configuring pass-through mirroring in registry?), we might want to phrase this like:

> is the recommended method for mirroring images onto a mirror registry.  This can be useful in several situations, including fully or partially disconnected environments.

Or something to make it clear that other situations (like "I just don't trust Quay to stay up/reachable/fast-for-my-moon-base/...") can happily use this process too.

File: modules/update-mirror-repository.adoc [hunk 2]
```diff
@@ -3,10 +3,8 @@
 // * updating/updating-restricted-network-cluster.adoc
 
 :_content-type: PROCEDURE
-[id="update-mirror-repository_{context}"]
```
Comment (wking) [inline modules/update-mirror-repository.adoc:6]: nit: should we preserve this old anchor to avoid breaking any previous linkers?  Or shift this anchor to a new location that mentions both the oc-mirror and `oc adm release mirror` options?

File: modules/update-mirror-repository.adoc [hunk 3]
```diff
@@ -3,10 +3,8 @@
...
-= Mirroring the {product-title} image repository
-
-Before you update a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. You can also use this procedure in unrestricted networks to ensure your clusters only use container images that have satisfied your organizational controls on external content.
+[id="update-mirror-repository_adm_release_mirror_{context}"]
+= Mirroring images using the adm release mirror command
```
Comment (wking) [inline modules/update-mirror-repository.adoc:8]: nit: this section will have installing `oc` as a prereq.  Not sure if it needs to get listed again here, or if we can inherit it from some wrapping context/module.

File: modules/update-mirror-repository-oc-mirror.adoc [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+.Procedure
+
+. Navigate to link:https://docs.openshift.com/container-platform/4.11/installing/disconnected_install/installing-mirroring-disconnected.html[Mirroring images for a disconnected installation using the oc-mirror plug-in].
+
+. Follow the instructions on that page up until but not including the section titled *Keeping your mirror registry content updated*.
```
Comment (wking) [inline modules/update-mirror-repository-oc-mirror.adoc:15]: Perhaps we need a third branch:

* Super easy: folks who have already set up oc-mirror, who only need [to keep their mirror up to date][1].
* Medium: folks who need to [set up oc-mirror for the first time][2], e.g. because they are pivoting from previously using `oc adm release mirror ...`.  These are the folks who would stop before `Keeping your mirror registry content updated`.
* Tedious: folks who keep using `oc adm release mirror ...`.

[1]: https://docs.openshift.com/container-platform/4.11/installing/disconnected_install/installing-mirroring-disconnected.html#updating-mirror-registry-content

File: modules/update-mirror-repository-oc-mirror.adoc (hunk 1 above)
Comment (skopacz1) [reply modules/update-mirror-repository-oc-mirror.adoc:9]: I'll rephrase the writing to what you suggested in your first comment

File: modules/update-mirror-repository.adoc (hunk 2 above)
Comment (skopacz1) [reply modules/update-mirror-repository.adoc:6]: This id has been removed from this module and added to the main "Mirroring the OpenShift Container Platform image repository" title in the assembly itself. So the link should still exist and work as before, and it should behave exactly the way you suggest in your second sentence.

However, I forgot to do the same thing for the OSUS section of this page, so I'll do that now. Thanks!

File: updating/updating-restricted-network-cluster.adoc [hunk 5]
```diff
@@ -38,11 +38,23 @@ include::modules/cli-installing-cli.adoc[leveloffset=+3]
...
 
-include::modules/machine-health-checks-pausing.adoc[leveloffset=+3]
+Before you update a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. You can also use this procedure in unrestricted networks to ensure your clusters only use container images that have satisfied your organizational controls on external content.
+
+include::modules/update-mirror-repository-oc-mirror.adoc[leveloffset=+3]
+
+[role="_additional-resources"]
+.Additional resources
+
+* xref:../installing/disconnected_install/installing-mirroring-disconnected.adoc[Mirroring images for a disconnected installation using the oc-mirror plug-in]
```
Comment (bergerhoffer) [inline updating/updating-restricted-network-cluster.adoc:53]: Good job making the extension .adoc, but you also need to add the anchor after it.

```suggestion
* xref:../installing/disconnected_install/installing-mirroring-disconnected.adoc#installing-mirroring-disconnected[Mirroring images for a disconnected installation using the oc-mirror plug-in]
```

FYI in case you aren't aware, don't click the "Commit suggestion" button that appears when someone does a suggested change like how I did here ^

File: updating/updating-restricted-network-cluster.adoc [hunk 6]
```diff
@@ -38,11 +38,23 @@ include::modules/cli-installing-cli.adoc[leveloffset=+3]
 
 // this file doesn't exist, so I'm including the one that should pick up more changes from Clayton's PR - modules/installation-adding-mirror-registry-pull-secret.adoc[leveloffset=+1]
 
-include::modules/installation-adding-registry-pull-secret.adoc[leveloffset=+2]
+include::modules/installation-adding-registry-pull-secret.adoc[leveloffset=+3]
 
-include::modules/update-mirror-repository.adoc[leveloffset=+2]
+[id="update-mirror-repository_{context}"]
```
Comment (bergerhoffer) [inline updating/updating-restricted-network-cluster.adoc:43]: Nice job including the ID above, but when you do IDs in assembly files, you don't add the `_{context}` piece. That piece is only used in module files.

File: modules/update-mirror-repository-oc-mirror.adoc [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+:_content-type: PROCEDURE
+[id="update-mirror-repository-oc-mirror_{context}"]
+= Mirroring resources using the oc-mirror plug-in
+
+Using the oc-mirror OpenShift CLI (`oc`) plug-in is the recommended method for mirroring images onto a mirror registry. This can be useful in several situations, including fully or partially disconnected environments.
+
+.Procedure
+
+. Navigate to _Mirroring images for a disconnected installation using the oc-mirror plug-in_.
```
Comment (bergerhoffer) [inline modules/update-mirror-repository-oc-mirror.adoc:13]: I'd recommend adding the "the <x> page" or "the <x> section" so that people know what you're talking about.

```suggestion
. Navigate to the _Mirroring images for a disconnected installation using the oc-mirror plug-in_ section of the documentation.
```

File: modules/update-mirror-repository-oc-mirror.adoc [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+. Follow the instructions on that page to mirror resources onto a mirror registry.
+
+** If you are using oc-mirror for the first time, follow the instructions on that page up until but not including the section titled *Keeping your mirror registry content updated*.
```
Comment (bergerhoffer) [inline modules/update-mirror-repository-oc-mirror.adoc:17]: I'm a little wary of using a section as where to stop, ahead of it. I did just add a new section to this assembly for 4.12. Luckily it was after "Keeping your...", but if it wasn't, it might have messed your instructions up.

It might be fine to leave it as is (though maybe risky). But I might play with calling out the exact section titles in ordered subbullets. Though I guess they do need to do the prerequisites and should read through the conceptual info, etc. so it might be a lot. I don't know, I'm just brainstorming here :) Give it some thought, but feel free to ignore

File: modules/update-mirror-repository.adoc [hunk 9]
```diff
@@ -3,10 +3,8 @@
 // * updating/updating-restricted-network-cluster.adoc
 
 :_content-type: PROCEDURE
-[id="update-mirror-repository_{context}"]
-= Mirroring the {product-title} image repository
-
-Before you update a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. You can also use this procedure in unrestricted networks to ensure your clusters only use container images that have satisfied your organizational controls on external content.
+[id="update-mirror-repository-adm-release-mirror_{context}"]
+= Mirroring images using the adm release mirror command
```
Comment (bergerhoffer) [inline modules/update-mirror-repository.adoc:7]: I'd include "oc" in this

```suggestion
= Mirroring images using the oc adm release mirror command
```

File: updating/updating-restricted-network-cluster.adoc [hunk 10]
```diff
@@ -123,7 +135,27 @@ include::modules/update-service-install-cli.adoc[leveloffset=+2]
...
 
-include::modules/update-service-mirror-release.adoc[leveloffset=+2]
+[id="update-service-mirror-release_{context}"]
+=== Mirroring the {product-title} image repository
+
+Before you update a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. You can also use this procedure in unrestricted networks to ensure your clusters only use container images that have satisfied your organizational controls on external content.
+
+//The module below is being used twice in this assembly, so this instance needs to have a unique context set in order for its ID to be unique. In the future, if and when the two main sections of this webpage are split into their own assemblies/pages, the context attributes below can be removed.
+
+:!context:
```
Comment (bergerhoffer) [inline updating/updating-restricted-network-cluster.adoc:145]: Although this probably works, I'm not sure if this something that we want to allow. 

I'd recommend checking in with @kalexand-rh on whether this is an acceptable workaround, or if we even have any other assemblies that try to include the same module twice. Seems like something we should probably not be doing, but there might be other places that do.

I know we talked about that this assembly should likely have been broken up into two assemblies. I know you mentioned it in your comment above about it maybe being done in the future, but did you give any thought in to doing that now with this update?

File: updating/updating-restricted-network-cluster.adoc [hunk 11]
```diff
@@ -123,7 +135,27 @@ include::modules/update-service-install-cli.adoc[leveloffset=+2]
...
+:context: OSUS-restricted-network-cluster
```
Comment (bergerhoffer) [inline updating/updating-restricted-network-cluster.adoc:146]: If you do end up keeping this setting/unsetting of context, then this OSUS should be lowercase.

File: modules/update-mirror-repository-oc-mirror.adoc [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+** If you have already used oc-mirror to mirror resources onto a registry, follow the instructions in the section titled *Keeping your mirror registry content updated*.
```
Comment (bergerhoffer) [inline modules/update-mirror-repository-oc-mirror.adoc:19]: We don't normally use bold to refer to section titles.

I typically use italics, but the jury is out on what we should use and there is a mix of italics and double quotes. But you should use one of those instead of bold.

**Review by bergerhoffer — COMMENTED** [review]
@skopacz1 really great work here! This is definitely complicated and our docs are unfortunately not well-suited for linking :roll_eyes: So we do our best. Left some suggestions and things to think about - let me know if you have any questions!

File: updating/updating-restricted-network-cluster.adoc (hunk 10 above)
Comment (kalexand-rh) [reply updating/updating-restricted-network-cluster.adoc:145]: I've never actually seen it happen, but it's something allowed in and described by [mod docs](https://redhat-documentation.github.io/modular-docs/#reusing-modules). (Search for "Module included multiple times in a single assembly")

It would be better to split it into two assemblies now, but it is legit to use the module twice.

Comment (kalexand-rh) [reply updating/updating-restricted-network-cluster.adoc:145]: I think it's the first time in the four years we've been doing mod docs that it's come up!

File: modules/update-mirror-repository-oc-mirror.adoc (hunk 8 above)
Comment (skopacz1) [reply modules/update-mirror-repository-oc-mirror.adoc:17]: I see what you mean about future updates potentially complicating these instructions. Part of me hopes that if a new section is created and it's not necessary to the basic workflow, it'll be added near the end of the page like you did with the new section. But I understand that this is not the best assumption to depend on.

I'll try out the bullet list and see how it looks. That said, I personally feel like if someone jams an optional/not-relevant-to-everyone section in the middle of these relevant sections, then in either case it might not be the best user experience to have to follow a bullet list on one page that doesn't exactly replicate the order of sections of the page being referenced. That's just my opinion though, I agree that this is a tricky situation with no great solution.

Comment (skopacz1) [reply modules/update-mirror-repository-oc-mirror.adoc:17]: I don't think this entirely addresses the concern you're mentioning here, but I'll change the wording so that it mentions the name of the section to stop **at**, not the one to stop **before**. Still risky but I hope it won't become a problem.

**Comment by skopacz1** [comment]
@shellyyang1989 PTAL, thank you!

File: modules/update-mirror-repository-oc-mirror.adoc [hunk 13]
```diff
@@ -0,0 +1,19 @@
+// Module included in the following assemblies:
+//
+// * updating/updating-restricted-network-cluster.adoc
```
Comment (shellyyang1989) [inline modules/update-mirror-repository-oc-mirror.adoc:3]: Should we indicate we have 2 alternative options to mirror images, either is fine `Mirroring resources using the oc-mirror plug-in` or `Mirroring images using the oc adm release mirror command`. Otherwise I'm worried some users might follow both sections to repeat the image mirror.

Comment (skopacz1) [reply modules/update-mirror-repository-oc-mirror.adoc:3]: I'll add more info to the paragraph that introduces these two options to make it clear to users that they should be making a choice between one or the other, and not doing both.

File: updating/updating-restricted-network-cluster.adoc [hunk 14]
```diff
@@ -38,11 +38,31 @@ include::modules/cli-installing-cli.adoc[leveloffset=+3]
...
-include::modules/machine-health-checks-pausing.adoc[leveloffset=+3]
+Before you update a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. You can also use this procedure in unrestricted networks to ensure your clusters only use container images that have satisfied your organizational controls on external content.
+
+There are two supported methods for mirroring images onto a mirror registry:
+
+* Using the oc-mirror OpenShift CLI (`oc`) plug-in (recommended)
+
+* Using the oc adm release mirror command
+
+Choose one of the supported options below.
```
Comment (skopacz1) [inline updating/updating-restricted-network-cluster.adoc:54]: @shellyyang1989 I added the info from lines 48-54 above in response to your feedback, does this look good to you?

Comment (shellyyang1989) [reply updating/updating-restricted-network-cluster.adoc:54]: yes, it looks good to me. Thanks!

**Review by JoeAldinger — COMMENTED** [review]
/remove-label peer-review-in-progress
/label peer-review-done

This looks pretty good. I left a few suggestions that I hope help.

File: updating/updating-restricted-network-cluster.adoc (hunk 14 above)
[the remark this answers is not in the document]
Comment (skopacz1) [reply updating/updating-restricted-network-cluster.adoc:54]: Do you think it would be acceptable if I were to remove this sentence entirely, and then replace the sentence on line 48 with the one below?

"Choose one of the two supported methods for mirroring images onto a mirror registry:"

[the remark this answers is not in the document]
Comment (JoeAldinger) [reply updating/updating-restricted-network-cluster.adoc:54]: I think it depends on whether or not this is a procedure or concept. I can see an argument for both options. As is, your heading title suggest procedure. If it is a procedure then I think it is fine as is, but you may want to rework some of the wording above to clarify. If it is a concept, I think the sentence would need to be reworked to introduce the procedures below.

File: updating/updating-restricted-network-cluster.adoc [hunk 15]
```diff
@@ -38,11 +38,31 @@ include::modules/cli-installing-cli.adoc[leveloffset=+3]
...
 
-include::modules/installation-adding-registry-pull-secret.adoc[leveloffset=+2]
+include::modules/installation-adding-registry-pull-secret.adoc[leveloffset=+3]
 
-include::modules/update-mirror-repository.adoc[leveloffset=+2]
+[id="update-mirror-repository_updating-restricted-network-cluster"]
+=== Mirroring the {product-title} image repository
 
-include::modules/machine-health-checks-pausing.adoc[leveloffset=+3]
+Before you update a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. You can also use this procedure in unrestricted networks to ensure your clusters only use container images that have satisfied your organizational controls on external content.
```
Comment (JoeAldinger) [inline updating/updating-restricted-network-cluster.adoc:46]:
```suggestion
You must, in a restricted network, mirror the required container images into that environment before you update a cluster on infrastructure that you provisioned. You can also mirror images in unrestricted networks to ensure your clusters only use container images that have satisfied your organizational controls on external content.
```

Comment (JoeAldinger) [reply updating/updating-restricted-network-cluster.adoc:46]: I _think_ another way to reword that first sentence might be: "You must mirror the required container images in a restricted network before you update a cluster on infrastructure that you provisioned." 

I was trying to reword to get to the action that a user must take earlier, which emphasizes that this is a procedure. I don't know if this works though.

File: updating/updating-restricted-network-cluster.adoc (hunk 14 above)
[the remark this answers is not in the document]
Comment (bergerhoffer) [reply updating/updating-restricted-network-cluster.adoc:54]: @skopacz1 instead of "below", you can use "following".

[the remark this answers is not in the document]
Comment (skopacz1) [reply updating/updating-restricted-network-cluster.adoc:54]: @JoeAldinger I see what you mean. The purpose of listing the two supported options was basically to create some place where I can explicitly mention that users are only supposed to do one of the two options and not both. So I think it's sort of like an intro to a procedure in that sense.

And thanks @bergerhoffer, I was hoping that I could use "following" instead but forgot to ask about it when I made my comment! I'll find some rewording that uses that instead.

File: modules/update-mirror-repository-oc-mirror.adoc [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+:_content-type: PROCEDURE
+[id="update-mirror-repository-oc-mirror_{context}"]
+= Mirroring resources using the oc-mirror plug-in
+
+Using the oc-mirror OpenShift CLI (`oc`) plug-in is the recommended method for mirroring images onto a mirror registry. This can be useful in several situations, including fully or partially disconnected environments.
```
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/update-mirror-repository-oc-mirror.adoc:9]: I'll change this excerpt around so it doesn't use"recommend", and list out the reasons it's recommended in the first place

File: updating/updating-restricted-network-cluster.adoc (hunk 15 above)
Comment (skopacz1) [reply updating/updating-restricted-network-cluster.adoc:46]: I'll play around to see if there's a way to reword this to get what you're going for. 

I don't think I'd want to split up "infrastructure that you provisioned" and "in a restricted network", because I think the first part only makes sense to include in this context when you specify that it's provisioned in a restricted network. Though maybe a clean looking reword won't be feasible while keeping this super clunky noun phrase the way it is. I'll test different options!
